### PR TITLE
linux: introduce io_uring support

### PR DESCRIPTION
**Pull request by bnoordhuis**
Add io_uring support for several asynchronous file operations:

- read, write
- fsync, fdatasync
- stat, fstat, lstat

io_uring is used when the kernel is new enough, otherwise libuv simply
falls back to the thread pool.

Performance looks great; an 8x increase in throughput has been observed.

This work was sponsored by ISC, the Internet Systems Consortium.

### CODE REVIEW

File: src/unix/linux.c
Comment (trevnorris) [inline src/unix/linux.c:673]: For metrics consistency will need to add `uv__metrics_update_idle_time(loop);` before this.

It would also be helpful if you could track the number of reqs that are processed and update `uv__metrics_inc_events()` as well. I don't immediately see how we would track `uv__metrics_inc_events_waiting()`.

I have this branch running locally. Will write tests to verify metrics with this specific case.

Comment (trevnorris) [reply src/unix/linux.c:673]: Actually, now I'm not sure about the call to `uv__metrics_update_idle_time()` because isn't it possible that the call to `epoll_pwait()` still idles? I'll just write the tests to verify.

Comment (bnoordhuis) [reply src/unix/linux.c:673]: > isn't it possible that the call to epoll_pwait() still idles?

Yes, until completion events become available in the ring buffer.

Comment (bnoordhuis) [reply src/unix/linux.c:673]: I think I can at least fix `metrics.loop_count` with the patch below. The remaining thorny issue is `metrics.events`.
<details>

```diff
diff --git a/src/unix/linux.c b/src/unix/linux.c
index dfe75c50..654bd1a0 100644
--- a/src/unix/linux.c
+++ b/src/unix/linux.c
@@ -700,6 +700,7 @@ void uv__io_poll(uv_loop_t* loop, int timeout) {
   sigset_t* sigmask;
   sigset_t sigset;
   uint64_t base;
+  int have_iou_events;
   int have_signals;
   int nevents;
   int count;
@@ -850,6 +851,7 @@ void uv__io_poll(uv_loop_t* loop, int timeout) {
       goto update_timeout;
     }

+    have_iou_events = 0;
     have_signals = 0;
     nevents = 0;

@@ -876,6 +878,7 @@ void uv__io_poll(uv_loop_t* loop, int timeout) {

       if (fd == iou->ringfd) {
         uv__poll_io_uring(loop, iou, fd, pe->events);
+        have_iou_events = 1;
         continue;
       }

@@ -950,6 +953,9 @@ void uv__io_poll(uv_loop_t* loop, int timeout) {
     loop->watchers[loop->nwatchers] = NULL;
     loop->watchers[loop->nwatchers + 1] = NULL;

+    if (have_iou_events != 0)
+      return;  /* Event loop should cycle now so don't poll again. */
+
     if (have_signals != 0)
       return;  /* Event loop should cycle now so don't poll again. */

```

</details>

**Comment by rektide** [comment]
Also potentially resolves or progresses #1947.

File: src/unix/linux.c
Comment (trevnorris) [reply src/unix/linux.c:778]: `loop_count` is taken care of somewhere else. That shouldn't be affected by any of the changes you made here.

Regarding `events`, I didn't understand the patch correctly. It should work if you changed `uv__poll_io_uring()` to return the number of times `cb()` is called, and change its call location to:
```c
nevents += uv__poll_io_uring(loop, iou, fd, pe->events);
```
and everything else will be taken care of since `nevents` is used to calculate both `events` and `events_waiting`.

Your change with `have_iou_events` will fix the `idle_time` calculation, but you'll still need to add `uv__metrics_update_idle_time(loop);` above the call here to `req->cb(req);`.

side note: I'm looking forward to node getting v1.45.0 in so I can show how useful these metrics are :)

**Comment by trevnorris** [comment]
Did some basic benchmarking, and it's giving me > 8x performance when reading small chunks out of `/dev/zero`. So great news on that side.

`valgrind` isn't happy about a few tests. Mostly with the error `Syscall param io_uring_enter(sigsz) contains uninitialised byte(s)`. Simplest repro I could make is this:
```c
#include <uv.h>

void on_read(uv_fs_t *req) { }

int main() {
  uv_fs_t read_req;
  uv_buf_t buffer;
  int fd;
  char bytes[64];

  buffer = uv_buf_init(bytes, 64);
  fd = uv_fs_open(NULL, &read_req, "/dev/zero", UV_FS_O_RDONLY, 0, NULL);
  uv_fs_read(uv_default_loop(), &read_req, fd, &buffer, 1, 0, on_read);
  uv_run(uv_default_loop(), UV_RUN_DEFAULT);
  uv_fs_close(NULL, &read_req, fd, NULL);

  uv_fs_req_cleanup(&read_req);
  uv_loop_close(uv_default_loop());
  return 0;
}
```

**Comment by bnoordhuis** [comment]
> personally I'd be fine with/love/prefer an io_uring that targeted more modern io_urings

Things may be heading that way out of necessity.

`IORING_SETUP_SQPOLL` requires linux v5.11 if you want to use it with non-fixed files (IIUC) but libuv can't do fixed files and without sqpoll it has to call `io_uring_enter(IORING_ENTER_GETEVENTS)` so often that any performance benefits are negated.

File: src/unix/linux.c
Comment (santigimeno) [inline src/unix/linux.c:575]: Are you planning to support something else in addition to `uv_fs_t` requests? If that's the case maybe have a generic function that takes a void* to fill the `user_data` field is useful, or naming it to reflect that's only for fs requests.

Comment (bnoordhuis) [reply src/unix/linux.c:575]: Only `uv_fs_t` for now and, since that's encoded in the function prototype, I don't really see the point of repeating it in the function name.

I've worked quite a bit with io_uring over the last year and I'm not that hopeful it's going to be useful for network i/o (or maybe I should say: will outperform epoll.)

To get the most mileage out you need to architect your code around io_uring's way of doing things but that's unfortunately not very compatible with libuv's way of doing things. I'll put together my notes and post them for discussion sometime.

File: src/unix/linux.c
Comment (bnoordhuis) [reply src/unix/linux.c:823]: I have it working now! I'll mark the discussion as resolved.

**Comment by bnoordhuis** [comment]
This is ready for review now. Linux v5.13 is the minimum required, otherwise libuv falls back to the thread pool.

File: src/unix/linux.c
Comment (santigimeno) [inline src/unix/linux.c:369]: Should this be documented?

File: test/test-threadpool-cancel.c
Comment (santigimeno) [inline test/test-threadpool-cancel.c:95]: Is it `IORING_OP_ASYNC_CANCEL` worth trying or as it doesn't work for disk IO requests it will never work?

File: src/unix/linux.c
Comment (santigimeno) [inline src/unix/linux.c:830]: Not 100% sure, but shouldn't this be increased by the number of completions just handled?

**Review by santigimeno — COMMENTED** [review]
Should we document this change in the `fs` docs?

File: src/unix/linux.c
Comment (bnoordhuis) [reply src/unix/linux.c:369]: I put this in for debugging purposes, in case people report bugs. I don't want to commit to making this a documented property just yet.

File: test/test-threadpool-cancel.c
Comment (bnoordhuis) [reply test/test-threadpool-cancel.c:95]: I opted to leave out cancellation for now. I do think it can be made to work but that's for another day (and pull request.)

File: src/unix/linux.c
Comment (bnoordhuis) [reply src/unix/linux.c:830]: metrics_info_check starts failing when I do that, cc @trevnorris.

**Comment by bnoordhuis** [comment]
> Should we document this change in the fs docs?

Yes. I thought it didn't say anything the implementation but it does. I'll do that tomorrow.

File: test/test-threadpool-cancel.c
Comment (santigimeno) [reply test/test-threadpool-cancel.c:95]: Got it. Are we ok leaving this support out without having linux users a "official" way to opt-out of iouring?

**Review by santigimeno — APPROVED** [review]
LGTM (once docs are updated). This is awesome, thanks

File: test/test-threadpool-cancel.c
Comment (bnoordhuis) [reply test/test-threadpool-cancel.c:95]: I'd argue it's not really a change in behavior because a libuv user has to handle UV_EBUSY from `uv_cancel()` anyway. I mean, we could change it to this:
```c
int uv_cancel(uv_req_t* req) { return UV_EBUSY; }
```
And it'd still be _pro forma_ backwards compatible. :-)

File: src/unix/linux.c
Comment (trevnorris) [reply src/unix/linux.c:824]: Thanks so much for taking care of that.

File: src/unix/linux.c
Comment (trevnorris) [reply src/unix/linux.c:831]: My initial thought is to treat uring events like normal events. Do that by incrementing `nevents`. That should also take care of returning early so there's no need for the additional `has_iou_events` variable. This diff passes all tests locally, but I'm going to write an additional test right now to double check handling of uring events:
```diff
diff --git a/src/unix/linux.c b/src/unix/linux.c
index d9eb9fec..e5f88ca1 100644
--- a/src/unix/linux.c
+++ b/src/unix/linux.c
@@ -782,6 +782,6 @@ static void uv__iou_fs_statx_post(uv_fs_t* req) {
 
-static void uv__poll_io_uring(uv_loop_t* loop,
-                              struct uv__iou* iou,
-                              int ringfd,
-                              int events) {
+static int uv__poll_io_uring(uv_loop_t* loop,
+                             struct uv__iou* iou,
+                             int ringfd,
+                             int events) {
   struct uv__io_uring_cqe* cqe;
@@ -793,2 +793,3 @@ static void uv__poll_io_uring(uv_loop_t* loop,
   uint32_t i;
+  uint32_t completed;
 
@@ -799,2 +800,3 @@ static void uv__poll_io_uring(uv_loop_t* loop,
   cqe = iou->cqe;
+  completed = 0;
 
@@ -824,2 +826,3 @@ static void uv__poll_io_uring(uv_loop_t* loop,
     req->cb(req);
+    completed++;
   }
@@ -830,3 +833,3 @@ static void uv__poll_io_uring(uv_loop_t* loop,
 
-  uv__metrics_inc_events(loop, 1);
+  return completed;
 }
@@ -855,3 +858,2 @@ void uv__io_poll(uv_loop_t* loop, int timeout) {
   uint64_t base;
-  int have_iou_events;
   int have_signals;
@@ -995,3 +997,2 @@ void uv__io_poll(uv_loop_t* loop, int timeout) {
 
-    have_iou_events = 0;
     have_signals = 0;
@@ -1021,4 +1022,3 @@ void uv__io_poll(uv_loop_t* loop, int timeout) {
       if (fd == iou->ringfd) {
-        uv__poll_io_uring(loop, iou, fd, pe->events);
-        have_iou_events = 1;
+        nevents += uv__poll_io_uring(loop, iou, fd, pe->events);
         continue;
@@ -1097,5 +1097,2 @@ void uv__io_poll(uv_loop_t* loop, int timeout) {
 
-    if (have_iou_events != 0)
-      return;  /* Event loop should cycle now so don't poll again. */
-
     if (have_signals != 0)
```

**Comment by trevnorris** [comment]
Still getting the following from valgrind:
```
fs_file_async

Syscall param io_uring_enter(sigsz) contains uninitialised byte(s)
   at 0x4993A3D: syscall (syscall.S:38)
   by 0x4F1366: uv__io_uring_enter (src/unix/linux.c:351)
   by 0x4F23A9: uv__iou_submit (src/unix/linux.c:644)
   by 0x4F24B5: uv__iou_fs_read_or_write (src/unix/linux.c:692)
   by 0x4DDAAC: uv_fs_write (src/unix/fs.c:2212)
   by 0x422F97: create_cb (test/test-fs.c:491)
   by 0x4DA38C: uv__fs_done (src/unix/fs.c:1779)
   by 0x4CDFD0: uv__work_done (src/threadpool.c:326)
   by 0x4D6433: uv__async_io (src/unix/async.c:176)
   by 0x4F2F95: uv__io_poll (src/unix/linux.c:1076)
   by 0x4D6B2D: uv_run (src/unix/core.c:417)
   by 0x42241C: run_test_fs_file_async (test/test-fs.c:915)
```

File: src/unix/linux.c
Comment (trevnorris) [reply src/unix/linux.c:831]: Just finished writing my first test, and appears that the way metrics are currently collected for these fs operations is broken. It's only incrementing by 1 for these all these fs events per loop instead of by the total number of fs events processed in the loop.

So, I have some debugging to do and more tests to write to see the extent of the problem. Not sure yet if this is specific to fs, or if it's a thread pool thing.

Here's the test I wrote that currently fails on `v1.x` but passes with the patch above in case you want to add it. Either way I'll be opening a PR to fix this: https://gist.github.com/trevnorris/bdaaf9e01cad30e1e509981fe0a8a8dd

Comment (trevnorris) [reply src/unix/linux.c:831]: @bnoordhuis in short, I need to write a PR to fix the metrics counter anyways. So you can choose to incorporate the patch I posted above or I'll be adding it as part of my fix.

**Review by trevnorris — APPROVED** [review]
LGTM other than the valgrind error. The current code you have for incrementing events is fine enough since it basically follows what the threadpool is doing. It's a bug, but not one you should need to worry about. I'll take care of that in a future PR.

File: src/unix/linux.c
Comment (trevnorris) [inline src/unix/linux.c:786]: Doesn't look like `ringfd` or `events` is being used. Is that intentional?

Comment (bnoordhuis) [reply src/unix/linux.c:786]: Yes (future extension for SQ notifications), but I'll remove them because I can see how it looks accidental.

**Comment by bnoordhuis** [comment]
FWIW, I don't get that warning with valgrind 3.18.1.

I'm 99% sure it's a valgrind bug because that `sigsz` it's complaining about is something that libuv doesn't use and doesn't exist in newer kernels.

On kernels where it was a thing libuv simply never calls `io_uring_enter()`.

**Comment by trevnorris** [comment]
@bnoordhuis I'm also running valgrind 3.18.1. Upon further testing, the error only appears when I compile with `clang` (I tried all versions 11 to 15). It doesn't happen with `gcc`. Think it's a bug in `clang` or just a strange interaction with `valgrind`?

**Comment by bnoordhuis** [comment]
@trevnorris figured it out while I was doing something completely different (isn't it always like that?), please see #3960. :-)

File: docs/src/fs.rst
Comment (bnoordhuis) [reply docs/src/fs.rst:17]: Just file operations for now. If you want to test without io_uring to compare, you can set `UV_USE_IO_URING=0` in the environment (but note that's undocumented and may go away in the future.)

Comment (bnoordhuis) [reply docs/src/fs.rst:17]: I don't plan on doing that but if it's a libuv bug, I'll fix it of course, and if you're hitting a kernel bug, I can probably add detection-and-fallback code. I'll need the kernel version, arch and faulting IORING_OP at the very least for that.

Comment (clason) [reply docs/src/fs.rst:17]: Affected kernels I know about:

* Arch Linux, kernel 6.3.1-arch1-1
* ChromeOS M113's 5.15.103-17409 inside of a Crostini container
* and possibly whatever GHA is currently using with ubuntu-latest

I'll have to see about tracing the IORING_OP calls for that (any advice or links for that?)

Comment (clason) [reply docs/src/fs.rst:17]: Here's the output of `perf top` on a hang:
```
  13.42%  [vdso]            [.] 0x00000000000006a5
   9.65%  [kernel]          [k] __put_user_nocheck_4
   7.75%  libc.so.6         [.] epoll_pwait
   7.32%  [kernel]          [k] __entry_text_start
   7.26%  [kernel]          [k] __x64_sys_epoll_pwait
   6.30%  [kernel]          [k] io_uring_poll
   6.25%  [kernel]          [k] __put_user_nocheck_8
   5.89%  nvim              [.] uv__io_poll
   5.06%  [kernel]          [k] do_epoll_wait
```

Comment (clason) [reply docs/src/fs.rst:17]: And here's a minimal reproducer (using luvit):
```lua
local function foo()
  --log("------------------------")
  local path ="/tmp/watch-test"
  for i = 1, 129 do
    local poll, err = uv.new_fs_poll()
    assert(poll and not err, err)
    --log("poll", i)
    poll:start(path, 2000, function(err1)
      if err1 == "ENOENT" then
        return
      end
      assert(not err1, err1)
    end)
  end
end

foo()
```
The 128th `poll:start()` blocks.

Comment (clason) [reply docs/src/fs.rst:17]: Working on a minimal example, but note that it's always exactly the 128th `poll:start()` call that is now blocking.

Comment (clason) [reply docs/src/fs.rst:17]: And here's a minimal reproducer (in Lua): https://github.com/neovim/neovim/issues/23595#issuecomment-1545648766

Comment (santigimeno) [reply docs/src/fs.rst:17]: It seems to me that the cq ring might have overflowed `IORING_SQ_CQ_OVERFLOW` and we might need to call `io_uring_enter()` to recover the events that didn't make it.

Comment (bnoordhuis) [reply docs/src/fs.rst:17]: Yes, I strongly suspect it's something along those lines. I'll open a pull request once I've managed to reproduce with just libuv.

(Or if you want to send one, of course go ahead.)

Comment (santigimeno) [reply docs/src/fs.rst:17]: @bnoordhuis I've given this a try in https://github.com/libuv/libuv/pull/3991. Let me know what you think. Thanks!

**Comment by aktau** [comment]
After Neovim merged the fix for overflow, I could start using it and it seems to work. But I see the following in `dmesg`:

> nvim: epoll_ctl support in io_uring is deprecated and will be removed in a future Linux kernel version.

Neovim does not appear to use `epoll_ctl(2)` itself, so this is likely libuv.

**Comment by GazHank** [comment]
Hi @bnoordhuis is there a simple way to test disabling this change per your previous suggestion of a `UV_USE_IO_URING=0` environment variable or compilation setting?

I'm hoping to rule out this change as the cause of a defect, so being able to quickly enable / disable the functionality would be very helpful

**Comment by bnoordhuis** [comment]
@GazHank `UV_USE_IO_URING=0 ` still exists for now. If you're using a 5.x kernel: we landed some follow-up fixes, not all of them released yet.

**Comment by GazHank** [comment]
thanks @bnoordhuis yes, I'm seeing some issues on ubuntu 22.04.2 LTS ( 5.19.0-46-generic ) 

So far I've managed to confirm that the issues only show up when upgrading from libuv 1.44 to 1.45, but setting an environment variable of `UV_USE_IO_URING=0` doesn't seem to make any difference.  Perhaps it is a different change from 1.45 that is causing problems, but I don't see any other likely causes within the notables changes list (as windows is entirely unaffected).

Please let me know if there is an existing issue you would like me to add info to, or if you would prefer that I raise a new issue entirely.  Of course if there are some draft changes you would like me to test to see if that solves the problem then please let me know, I'd be happy to help test the changes.

**Comment by bnoordhuis** [comment]
You can try with the HEAD of the v1.x branch. If that doesn't fix it, it'd be helpful if you could run `git bisect` on the commits between 1.44 and 1.45 and open an issue with as much details as possible.

**Comment by GazHank** [comment]
Thanks @bnoordhuis  Unfortunately I can confirm that it is this commit that causes the break :-(

Additionally, setting `UV_USE_IO_URING=0` doesn't seem to make any difference, however amending the kernel version specified within linux.c file to a higher version than I'm running does fix/workaround the issue

https://github.com/libuv/libuv/blob/d09441ca03e399fe641da88624c8ea0476967187/src/unix/linux.c#L431-L443

As such, is it perhaps that `UV_USE_IO_URING=1` can be use to enable the feature, but `UV_USE_IO_URING=0` cannot be used to disable it?

I'll raise a new issues and try to include as much info as I can

**Comment by bnoordhuis** [comment]
`UV_USE_IO_URING=0` is a hard disable. You should be able to verify with strace that no io_uring system calls are made.

**Comment by wswsmao** [comment]
> [trevnorris](/trevnorris)



> Did some basic benchmarking, and it's giving me > 8x performance when reading small chunks out of `/dev/zero`. So great news on that side.
> 
> `valgrind` isn't happy about a few tests. Mostly with the error `Syscall param io_uring_enter(sigsz) contains uninitialised byte(s)`. Simplest repro I could make is this:
> 
> ```c
> #include <uv.h>
> 
> void on_read(uv_fs_t *req) { }
> 
> int main() {
>   uv_fs_t read_req;
>   uv_buf_t buffer;
>   int fd;
>   char bytes[64];
> 
>   buffer = uv_buf_init(bytes, 64);
>   fd = uv_fs_open(NULL, &read_req, "/dev/zero", UV_FS_O_RDONLY, 0, NULL);
>   uv_fs_read(uv_default_loop(), &read_req, fd, &buffer, 1, 0, on_read);
>   uv_run(uv_default_loop(), UV_RUN_DEFAULT);
>   uv_fs_close(NULL, &read_req, fd, NULL);
> 
>   uv_fs_req_cleanup(&read_req);
>   uv_loop_close(uv_default_loop());
>   return 0;
> }
> ```

Hi, @trevnorris , I want to evaluate the performance improvements brought by this feature. I was wouldering is there any performance  benchmarking available？

**Comment by wswsmao** [comment]
> [bnoordhuis](/bnoordhuis)

Hi, @bnoordhuis , I was wouldering is there any performance benchmarking available？

**Comment by wswsmao** [comment]
> > [trevnorris](/trevnorris)
> 
> > Did some basic benchmarking, and it's giving me > 8x performance when reading small chunks out of `/dev/zero`. So great news on that side.
> > `valgrind` isn't happy about a few tests. Mostly with the error `Syscall param io_uring_enter(sigsz) contains uninitialised byte(s)`. Simplest repro I could make is this:
> > ```c
> > #include <uv.h>
> > 
> > void on_read(uv_fs_t *req) { }
> > 
> > int main() {
> >   uv_fs_t read_req;
> >   uv_buf_t buffer;
> >   int fd;
> >   char bytes[64];
> > 
> >   buffer = uv_buf_init(bytes, 64);
> >   fd = uv_fs_open(NULL, &read_req, "/dev/zero", UV_FS_O_RDONLY, 0, NULL);
> >   uv_fs_read(uv_default_loop(), &read_req, fd, &buffer, 1, 0, on_read);
> >   uv_run(uv_default_loop(), UV_RUN_DEFAULT);
> >   uv_fs_close(NULL, &read_req, fd, NULL);
> > 
> >   uv_fs_req_cleanup(&read_req);
> >   uv_loop_close(uv_default_loop());
> >   return 0;
> > }
> > ```
> 
> Hi, @trevnorris , I want to evaluate the performance improvements brought by this feature. I was wouldering is there any performance benchmarking available？

Hi, @bnoordhuis , I was wouldering is there any performance benchmarking available？

File: src/unix/linux.c
Comment (heiher) [inline src/unix/linux.c:637]: These lines of code are equivalent to:

```c
  atomic_thread_fence(memory_order_release);
  iou->sqtail = *iou->sqtail + 1;
  flags = iou->sqflags;
  atomic_thread_fence(memory_order_acquire);
```

It seems impossible to prevent the load of `sqflags` from being reordered before the store to `sqtail`. e.g.

```c
  flags = iou->sqflags;
  ...
  atomic_thread_fence(memory_order_release);
  iou->sqtail = *iou->sqtail + 1;
  atomic_thread_fence(memory_order_acquire);
```

I suggest the following change:

```c
  atomic_store_explicit((_Atomic uint32_t*) iou->sqtail,
                        *iou->sqtail + 1,
                        memory_order_release);
  atomic_thread_fence(memory_order_seq_cst);

  if (iou->sqflags & UV__IORING_SQ_NEED_WAKEUP)
```

equivalent to:

```c
  atomic_thread_fence(memory_order_release);
  iou->sqtail = *iou->sqtail + 1;
  atomic_thread_fence(memory_order_seq_cst);

  if (iou->sqflags & UV__IORING_SQ_NEED_WAKEUP)
```

Comment (heiher) [reply src/unix/linux.c:637]: Ref: https://github.com/torvalds/linux/blob/691d401c7e0e5ea34ac6f8151bc0696db1b2500a/io_uring/io_uring.c#L25-L28

```
 * When using the SQ poll thread (IORING_SETUP_SQPOLL), the application
 * needs to check the SQ flags for IORING_SQ_NEED_WAKEUP *after*
 * updating the SQ tail; a full memory barrier smp_mb() is needed
 * between.
```

**Pull request merged**
